pkg/user/tx_client.go

```
@@ -415,6 +441,54 @@ func (client *TxClient) broadcastTx(ctx context.Context, txBytes []byte, signer
 	return resp.TxResponse, nil
 }
 
+// broadcastMulti broadcasts the transaction to multiple connections concurrently
+// and returns the response from the first successful broadcast.
+func (client *TxClient) broadcastMulti(ctx context.Context, txBytes []byte, signer string) (*sdktypes.TxResponse, error) {
+	respCh := make(chan *sdktypes.TxResponse, 1)
+	errCh := make(chan error, len(client.conns))
+
+	g, childCtx := errgroup.WithContext(ctx)
+
+	for _, conn := range client.conns {
+		g.Go(func() error {
+			resp, err := client.broadcastTx(childCtx, conn, txBytes, signer)
+			if err != nil {
+				errCh <- err
+				return nil
+			}
+
+			select {
+			case respCh <- resp:
+			case <-childCtx.Done():
+			}
+			return nil
+		})
+	}
+
+	groupErr := g.Wait()
```

here we wait for *all* requests to return, we do not cancel other in-flight requests once a success response is read. This logic is missing.

---

no, as the first one returns the group is cancelled and they return quickly

---

@yarikbratashchuk Where do explicitly cancel out the other in-flight routines here? `g.Wait()` blocks until *all* routines launched in that loop you have above have exited. You do not cancel the childCtx here, so I'm confused where the quick return happens?  

Also I don't understand what you mean by respCh being buffered - the buffer is there only to be able for sender to write to ch without being blocked on reader not being ready to read yet, so I'm confused what you mean by "we don't wait before it's read" ?

---

@renaynay on success we return err which cancels childCtx withing the errgroup.

---

Okay I see what the misunderstanding is here. This was quite confusing logic at first and not very explicit because you opted to use errgroup to take advantage of fail fast option, but inversely for a success case 😅  You are returning `context.Canceled` which is a non-nil error upon success case and signalling to errgroup to cancel childCtx which will be caught in other in-flight routines via select -- yes the logic technically works, but is difficult to read as errgroups should rather be used for operations where you want to parallelise some work but one error from one worker should cancel out operations on rest of groups + better err handling for that generally. 

Here, I would've opted for sync.WaitGroup and explicitly done the cancellation of ctx for other in-flight routines so it doesn't read like a success case --> err (then having to remember that err group cancels childCtx upon first err, whereas operations that did not finish in timeout return nil error rather than ctx Err). 

Tl;dr both options work - sync.WaitGroup would read better here IMO but i'm fine with keeping errgroup as long as the ctx cancellation is more explicit.